### ISSUE TITLE
gazelle vendor support

### ISSUE THREAD

**Issue by hochhaus** [bazelbuild/rules_go#114]
Is vendoring automatically supported when using gazelle/`new_go_repository()`?

As a specific example, when using `new_go_repository()` with a repository that uses vendoring (eg: https://github.com/rsc/letsencrypt) I see that gazelle creates BUILD files for the vendored repositories.

For example, adding the following to my WORKSPACE:

```
new_go_repository(
    name = "io_rsc_letsencrypt",
    commit = "a18c646c3d0772313b7b53c78847bb3eb6463f0b",
    importpath = "rsc.io/letsencrypt",
)
```

generates the following BUILD files:

```
ahochhaus@ahochhaus-pc:~/ll$ find -L bazel-* -name BUILD | grep letsen
bazel-ll/external/io_rsc_letsencrypt/BUILD
bazel-ll/external/io_rsc_letsencrypt/vendor/github.com/xenolf/lego/acme/BUILD
```

However when I build, the vendored repository (`com_github_xenolf_lego`) can't be found.

**Comment by pcj** [comment]
(chiming in)

There would have to be a `WORKSPACE` file written at `$(bazel info output_base)/external/com_github_xenolf_lego` for this to work.  Does that file exist?  Otherwise I'd think it one would have to refer to it as `@io_rsc_letsencrypt//vendor/github.com/xenolf/lego/acme:go_default_library`.

**Comment by pmbethe09** [comment]
So running gazelle creates deps by using a consistent conversion string
from Go import-path to a workspace-name.
That is true both for manual local gazelle, and gazelle underneath a
new_go_repository.

However in both cases that assumes you have all of the right things in your
WORKSPACE file.

Presumably what we need is either a stand-alone "workzelle", or something
running as part of local invocation of 'gazelle -workspace=fix ...'
This would collect all of the remote-repo deps as referenced by gazelle,
and add them to your WORKSPACE, as well as look at the transitive
dependencies.
Note: this is because bazel require a flattening of all transitive
dependencies into your local WORKSPACE file -- it doesn't read WORKSPACE
files read from any remote repos.

I have started working on such a thing, but want to touch base with @yugui
to plan

**Comment by yugui** [comment]
> I have started working on such a thing, but want to touch base with @yugui to plan

Both of `workzelle` and `-workspace=fix` sound good for me. But probably `workzelle` is simpler.

Also, I just want to share another point in my mind with you -- it might be related to this topic.  Some users (e.g. my team) uses both of `new_go_repository` rule and `vendor/` directory. So they would need a way to customize the mapping from the importpath to label in `gazelle`.

**Comment by hochhaus** [comment]
Thanks for the work on gazelle and all the great feedback!

As best I can tell, none of these proposals match the semantics of the go tool. My initial bug report was not clear about what I am looking for.

The problem is that multiple repositories can all depend on different versions of `com_github_xenolf_lego` (as that package could be vendored at different versions in two or more libraries).

Instead of the proposed behavior, it seems better to me that the vendored copy of `com_github_xenolf_lego` in `io_rsc_letsencrypt` actually be named something like `@io_rsc_letsencrypt_vendored_com_github_xenolf_lego`. Also the build files generated for `io_rsc_letsencrypt` should automatically depend on the vendored package.

However, at that point, why can't the dependencies in `bazel-ll/external/io_rsc_letsencrypt/BUILD` automatically reference `bazel-ll/external/io_rsc_letsencrypt/vendor/github.com/xenolf/lego/acme/BUILD` hence eliminating the need for an extra WORKSPACE entry for the vendored package all together.

Does this behavior seem reasonable?

**Comment by pmbethe09** [comment]
On Fri, Sep 23, 2016 at 2:27 AM, Andy Hochhaus <notifications@github.com>
wrote:

> Thanks for the work on gazelle and all the great feedback!
>
> As best I can tell, none of these proposals match the semantics of the go
> tool. My initial bug report was not clear about what I am looking for.
>
> The problem is that multiple repositories can all depend on different
> versions of com_github_xenolf_lego (as that package could be vendored at
> different versions in two or more libraries).
>
> Instead of the proposed behavior, it seems better to me that the vendored
> copy of com_github_xenolf_lego in io_rsc_letsencrypt actually be named
> something like @io_rsc_letsencrypt_vendored_com_github_xenolf_lego. Also
> the build files generated for io_rsc_letsencrypt should automatically
> depend on the vendored package.
>
So here by vendored, you mean checked into a vendored directory, rather
than pinned to a particular version?

For the remote-vendored diamond dependency problem, bazel wants you to
choose the winner by flattening your workspace.


> However, at that point, why can't the dependencies in
> bazel-ll/external/io_rsc_letsencrypt/BUILD automatically reference
> bazel-ll/external/io_rsc_letsencrypt/vendor/github.com/
> xenolf/lego/acme/BUILD hence eliminating the need for an extra WORKSPACE
> entry for the vendored package all together.
>
Some people want that, others prefer a flattened WORKSPACE file that
clearly identifies all direct and indirect dependencies, so that there are
no surprises.

**Comment by hochhaus** [comment]
> So here by vendored, you mean checked into a vendored directory, rather than pinned to a particular version?

Yes, for example [rsc/letsencrypt/vendor/github.com/xenolf/lego](https://github.com/rsc/letsencrypt/tree/master/vendor/github.com/xenolf/lego).

> For the remote-vendored diamond dependency problem, bazel wants you to
choose the winner by flattening your workspace.

Unfortunately, you can't always "pick a single winner" as the API could have changed between library versions. Since GOVENDOREXPERIMENT became the default in Go 1.6 the standard go build tools assume that multiple different versions of the same library can be compiled into a single binary. I think that if `rules_go` does not follow a similar convention it will be harder to get general interoperability with the rest of the open source go ecosystem.

In the case of `io_rsc_letsencrypt`, it really does depend on an outdated API of `com_github_xenolf_lego` (specifically `TLSSNI01ChallengeCertDomain()`). Any other use of `com_github_xenolf_lego` which uses the new API (`TLSSNI01ChallengeCert()`) in my project (or a library used by my project) will mean that no single version of `com_github_xenolf_lego` can be entered into the `WORKSPACE` file.

> Some people want that, others prefer a flattened WORKSPACE file that
clearly identifies all direct and indirect dependencies, so that there are
no surprises.

I certainly see value in both models. For example, a downside of using the `vendor/` directory model that I propose is that it is less predictable and leads to binary bloat. However, a downside of forcing a single version of each dep to be entered into the WORKSPACE file is that it is less interoperable with the rest of the Go ecosystem and may require code changes or manually written BUILD files.

At a minimum could we make this behavior configurable via an option?

**Comment by pcj** [comment]
I agree, the whole point of vendoring is that you can maintain several different versions of the same library in the same WORKSPACE, so it does not seem right to hoist a vendored name into the global at-namespace.  

I'm not in love with the `_vendored` part of @io_rsc_letsencrypt_vendored_com_github_xenolf_lego (name is long enough, isn't it already implied?) but I completely agree with the design.

**Comment by hochhaus** [comment]
> I'm not in love with the _vendored part of @io_rsc_letsencrypt_vendored_com_github_xenolf_lego (name is long enough, isn't it already implied?) but I completely agree with the design.

Good point. I'm happy with any name. I was just trying to propose the concept with a concrete example.

**Comment by pmbethe09** [comment]
I re-read this and realized that what is really needed for consistency with the Go tool, is gazelle generating dependencies to things that are vendored, rather than to external repos.
So in this case rather than a dep on "@com_github_xenolf_lego//acme:go_default_library" it should be as @pcj wrote above, to "io_rsc_letsencrypt//vendor/github.com/xenolf/lego/acme:go_default_library"

Probably:
* Add -ignore_vendored flag to gazelle (where false will become default)
** if true, assumes all deps are remote
** if false reads a //vendor directory
* Add ignore_vendored attribute to new_go_repository, which passes through to the gazelle invocations
** allows users to make WORKSPACE decision to avoid vendoring, but the default of False means "do what 'go tool' would have done"

**Comment by mikedanese** [comment]
To workaround no support for vendor/ in gazelle, the kubernetes project is using https://github.com/mikedanese/gazel

**Comment by pmbethe09** [comment]
What is your gazel?  always use vendored? or are there other features?

**Comment by mikedanese** [comment]
It manages most of the BUILD files in kubernetes. It produces a single very big BUILD file for vendor https://github.com/kubernetes/kubernetes/blob/master/vendor/BUILD. We don't use go_library remote repository rules yet and may never. As far as I could tell, gazelle assumes this. It also allows for partially managed BUILD files. With gazelle it's all or none. There may be other differences as well.

**Comment by pcj** [comment]
@mikedanese, can you add a README to your repository that shows how to use it?

**Comment by mikedanese** [comment]
@pcj added a README

**Comment by pcj** [comment]
@mikedanese Nit: will render better as `README.md`

**Comment by twpayne** [comment]
#227 contains a partial fix, I think.

**Comment by kerinin** [comment]
I think what's missing after #227 is the ability to generate BUILD files in `vendor` with the correct go prefix

**Comment by mikedanese** [comment]
Preferably a single BUILD file in the root of vendor/

**Comment by MarkusTeufelberger** [comment]
This would be pattern 2 in https://github.com/bazelbuild/rules_go/issues/16#issuecomment-216010843, right?

As far as i understand it, gazelle only implements patterns 1 and 3 so far, while the gazel tool used in Kubernetes implements pattern 2. It would be nice if gazelle also supported that pattern.

**Comment by jayconrod** [comment]
Close old Gazelle issues. If Gazelle isn't doing something you want, please open a new issue at https://github.com/bazelbuild/bazel-gazelle.

This should be working now. Gazelle will resolve dependencies against existing libraries in vendor in either external or vendored mode.
